File: leetcode/28.找出字符串中第一个匹配项的下标/28.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        next = [0] * len(needle)
        self.getNext(next, needle)
        i, j = 0, 0
        while i < len(haystack):
            if haystack[i] == needle[j]:
                i += 1
                j += 1
            elif j > 0:
                j = next[j - 1]
            else:
                i += 1
            if j == len(needle):
                return i - j
        return -1

    def getNext(self, next, s:str):
        prefix_len, i = 0, 1
        while i < len(s):
            if s[prefix_len] == s[i]:
                prefix_len += 1
                next[i] = prefix_len
                i += 1
            else:
                if prefix_len == 0:
                    next[i] = prefix_len
                    i += 1
                else:
                    prefix_len = next[prefix_len - 1]

File: leetcode/28.找出字符串中第一个匹配项的下标/28.java
class Solution {
    public int strStr(String haystack, String needle) {
        int[] next = new int[needle.length()];
        next[0] = 0;
        getNext(next, needle);
        int  i = 0, j = 0; 
        while(i < haystack.length()){
            if(haystack.charAt(i) == needle.charAt(j)){
                i++;
                j++;
            }else if(j > 0){
                j = next[j - 1];
            }else{
                i++;
            }
            if(j == needle.length()) return i - j;
        }
        return -1;
    }
    public void getNext(int[] next, String s){
        int prefix_len = 0, i = 1;
        while(i < s.length()){
            if(s.charAt(i) == s.charAt(prefix_len)){
                next[i++] = ++prefix_len;
            }else{
                if(prefix_len == 0){
                    next[i++] = 0;
                }else{
                    prefix_len = next[prefix_len - 1];
                }
            }
        }
    }
}